scripts/main.py

```diff
@@ -22,85 +24,130 @@ def ui(self, is_img2img):
+                        label=f"Backbone {block_n} Scale",
+                        minimum=-1,
+                        maximum=3,
+                        value=default_block_info.backbone_factor,
+                    ))
+                    default_block_info.backbone_factor = block_flat_components[-1].value
+
+                    block_flat_components.append(gr.Slider(
+                        label=f"Skip {block_n} Scale",
+                        minimum=-1,
+                        maximum=3,
+                        value=default_block_info.skip_factor,
+                    ))
+                    default_block_info.skip_factor = block_flat_components[-1].value
+
+                    with gr.Row():
+                        block_flat_components.append(gr.Slider(
+                            label=f"Backbone {block_n} Offset",
+                            minimum=0,
+                            maximum=1,
+                            value=default_block_info.backbone_offset,
+                        ))
+                        default_block_info.backbone_offset = block_flat_components[-1].value
+
+                        block_flat_components.append(gr.Slider(
+                            label=f"Backbone {block_n} Width",
+                            minimum=0,
+                            maximum=1,
+                            value=default_block_info.backbone_width,
+                        ))
+                        default_block_info.backbone_width = block_flat_components[-1].value
+
+                    flat_components.extend(block_flat_components)
 
         reset_to_defaults.click(
-            fn=lambda: (1.2, 0.9, 0, 0.5, 1.4, 0.2, 0, 0.5),
-            outputs=[b0, s0, o0, w0, b1, s1, o1, w1],
+            fn=lambda: [
+                v
+                for block_info in default_block_infos
+                for v in block_info.to_dict().values()
+            ],
+            outputs=flat_components,
+        )
+
+        infotext_component = gr.HTML(visible=False, interactive=False)
+
+        infotext_component.change(
+            fn=self.on_infotext_update,
+            inputs=[infotext_component],
+            outputs=[infotext_component, *flat_components],
         )
 
-        return enabled, b0, s0, o0, w0, b1, s1, o1, w1
+        self.infotext_fields = [
+            (infotext_component, "FreeU"),
+        ]
+        self.paste_field_names = ["FreeU"]
+
+        return enabled, *flat_components
+
+    def on_infotext_update(self, infotext):
+        if not infotext:
+            return (gr.skip(),) * 9
+
+        params = json.loads(infotext)
+        block_infos = [
+            global_state.BlockInfo(*block_info.values())
+            for block_info in params["block_infos"]
+        ]
+
+        return (
+            gr.update(value=""),
+            *(
+                gr.update(value=v)
+                for block_info in block_infos
+                for v in block_info.to_dict().values()
+            )
+        )
```

Comment:
This does not scale. Adding a new slider breaks older png info.

Reply:
The way it should work is that if a slider is absent in the PNG info, we set that slider to the default setting (so that the gen can be reproduced). Conversely, new sliders must have their default value such that gens aren't affected by the presence of that slider.

Reply:
The thing is, the default value will depend on the model type in the future, so there won't really be a stable "default" to rely on. This would work if the model was changed along with png info, but this setting can be disabled.

Reply:
It would be really cool if we could approximate the best values dynamically during inference also, although reproducing generations here would be a little bit more tricky.

Reply:
If know that the generation is prior to the introduction of a new slider, then we know that slider's state (it is whatever slider value causes no effect)

Reply:
yeah let's go with that